File: go/models/validate/validate.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Script to ensure datastore definitions in Python and Go are compatible."""
import datetime
import subprocess
import sys

from google.cloud import ndb

import osv.tests
from osv import Vulnerability


def main() -> int:
  # Populate the examples from Python
  print('(Python) Putting Vulnerability')
  Vulnerability(
      id='CVE-123-456',
      source_id='test:path/to/CVE-123-456.json',
      modified=datetime.datetime(2025, 1, 2, 3, 4, 5, tzinfo=datetime.UTC),
      is_withdrawn=False,
      modified_raw=datetime.datetime(2025, 1, 1, 1, 1, 1, tzinfo=datetime.UTC),
      alias_raw=['OSV-123-456', 'TEST-123-456'],
      related_raw=['CVE-000-000', 'CVE-111-111'],
      upstream_raw=['CVE-123-000', 'OSV-123-000'],
  ).put()

  # Run Go program to read the Python-created entities in Go.
  # And write Go entities.
  result = subprocess.run(['go', 'run', './validate.go'], check=False)
  if result.returncode != 0:
    return result.returncode

  # Read the Go-created entities in Python.
  print('(Python) Getting Vulnerability')
  if Vulnerability.get_by_id('CVE-987-654') is None:
    return 1
  return 0


if __name__ == '__main__':
  with osv.tests.datastore_emulator(), ndb.Client().context():
    ret = main()
  sys.exit(ret)

File: go/models/internal/validate/validate.go
package main

import (
	"context"
	"fmt"
	"os"
	"time"

	"cloud.google.com/go/datastore"
	"github.com/google/osv.dev/go/models"
)

func main() {
	ctx := context.Background()
	client, err := datastore.NewClient(ctx, os.Getenv("GOOGLE_CLOUD_PROJECT"))
	if err != nil {
		fmt.Printf("(Go) Failed creating Datastore client: %v\n", err)
		os.Exit(1)
	}

	readRecords(ctx, client)
	writeRecords(ctx, client)
}

func readRecords(ctx context.Context, client *datastore.Client) {
	fmt.Println("(Go) Getting Vulnerability")
	key := datastore.NameKey("Vulnerability", "CVE-123-456", nil)
	var vulnerability models.Vulnerability
	if err := client.Get(ctx, key, &vulnerability); err != nil {
		fmt.Printf("(Go) Failed getting Vulnerability: %v\n", err)
		os.Exit(1)
	}
}

func writeRecords(ctx context.Context, client *datastore.Client) {
	fmt.Println("(Go) Writing Vulnerability")
	key := datastore.NameKey("Vulnerability", "CVE-987-654", nil)
	vulnerability := models.Vulnerability{
		SourceID:    "test:path/to/CVE-987-654",
		Modified:    time.Date(2025, time.December, 31, 23, 59, 59, 0, time.UTC),
		IsWithdrawn: false,
		ModifiedRaw: time.Date(2025, time.December, 1, 23, 59, 59, 0, time.UTC),
		AliasRaw:    []string{"OSV-987-654", "TEST-987-654"},
		RelatedRaw:  []string{"CVE-999-999"},
		UpstreamRaw: []string{"CVE-987-000", "OSV-987-000"},
	}
	if _, err := client.Put(ctx, key, &vulnerability); err != nil {
		fmt.Printf("(Go) Failed writing Vulnerability %v: %v\n", key, err)
		os.Exit(1)
	}
}
